common/app/Map/redux/utils.js

```diff
@@ -157,3 +157,75 @@ export function collapseAllPanels(tree) {
 export function expandAllPanels(tree) {
   return toggleAllPanels(tree, true);
 }
+
+// synchronise
+// updatePath(
+//   tree: MapUi,
+//   name: String,
+//   update(MapUi|Node) => MapUi|Node
+// ) => MapUi
+export function updatePath(tree, name, pathUpdater) {
+  const path = [];
+  let pathFound = false;
+
+  const isInPath = node => !!path.find(name => name === node.name);
+
+  const traverseMap = (tree, update) => {
```

Comment:
There is already a traverse map function. What is the requirement for this new function that the other does not meet?

Reply:
@BerkeleyTrue current traverse map function doesn't allow you track parent nodes, i.e. path to the node, so we have to store nodes in path before going deeper.

Reply:
@BerkeleyTrue changed this and moved meta creator to map as well.

Reply:
@vbelolapotkov can you explain what the path is and why you need to track it? The traversal is a depth first search algorithm. This should be sufficient to find the current challenge and open the map to expose that challenge.

Reply:
@BerkeleyTrue to open map I have to set isOpen = true on nodes with children in path (superBlock > block > challenge). But I have only challenge name as a starting point and do not have any link from child to parent (challenge > block > superBlock). So if I find challenge node by name, I have no way to update it's parents (block, superBlock). That is why I'm saving parent in path any time I go deeper with traversal.

Reply:
Why do you need the full path?

Reply:
With full path I can check if node is a parent of challenge node and update those nodes when going out of the recursion. 
Do you have an idea on how to do this with traverse tree?

Reply:
I believe it may be sufficient to to check for open children in the node and open the current node if any of it's children are open.

Reply:
This involves double check of each node. I'd prefer to avoid it.

Reply:
> This involves double check of each node.

How do you figure this?

Reply:
`update` is called on each child before it's called on a parent node. When it is called on a parent node you have to look through the children again to find out if any of them is opened. 
So you have to iterate twice: first iteration during a traverse, second iteration in update.

Reply:
I see. Thanks for clarifying